### PR TITLE
Face mask classifier

### PR DESCRIPTION
**Pull request by mrn-mln**
A structure for a classifier is added to the code. The classifier works with an openpifpaf detector at x86 devices.

### CODE REVIEW

File: libs/detectors/x86/openpifpaf.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            xmax = int(min(x_max + .15 * w, self.w))
+            ymin = int(max(y_min - .2 * h, 0))
+            ymax = int(min(y_max + .05 * h, self.h))
+            bbox_dict = {"id": "1-" + str(i), "bbox": [ymin / self.h, xmin / self.w, ymax / self.h, xmax / self.w],
+                    "score": 0.9, "face": None}
+            # extracting face bounding box
+            if np.all(pred[[0, 1, 2, 5, 6], -1] > 0.15):
+                x_min_face = int(pred[6, 0])
+                x_max_face = int(pred[5, 0])
+                y_max_face = int((pred[5, 1] + pred.data[6, 1]) / 2)
```
Comment (alpha-carinae29) [inline libs/detectors/x86/openpifpaf.py:81]: should change to `y_max_face = int((pred[5, 1] + pred[6, 1]) / 2)`

File: libs/detectors/x86/openpifpaf.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                y_eyes = int((pred[1, 1] + pred.data[2, 1]) / 2)
```
Comment (alpha-carinae29) [inline libs/detectors/x86/openpifpaf.py:82]: change to `y_eyes = int((pred[1, 1] + pred[2, 1]) / 2)`

File: libs/detectors/x86/openpifpaf.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                y_min_face = 2 * y_eyes - y_max_face
+                if (y_max_face - y_min_face > 0) and (x_max_face - x_min_face > 0):
+                    h_crop = y_max_face - y_min_face
+                    x_min_face = int(max(0, x_min_face - 0.1 * h_crop))
+                    y_min_face = int(max(0, y_min_face - 0.1 * h_crop))
+                    x_max_face = int(min(w, x_min_face + 1.1 * h_crop))
```
Comment (alpha-carinae29) [inline libs/detectors/x86/openpifpaf.py:88]: change to `x_max_face = int(min(self.w, x_min_face + 1.1 * h_crop))`

File: libs/detectors/x86/openpifpaf.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    y_max_face = int(min(h, y_min_face + 1.1 * h_crop))
```
Comment (alpha-carinae29) [inline libs/detectors/x86/openpifpaf.py:89]: change to `y_max_face = int(min(self.h, y_min_face + 1.1 * h_crop))`

**Comment by mats-claassen** [comment]
Hi @mrn-mln could you rebase this on top of master?

**Comment by mats-claassen** [comment]
Hey @mrn-mln I saw you got a PR merged into the Neuralet repo [here](https://github.com/neuralet/neuralet/pull/133). What is the status of this one? Is it still valid? Does it have to be updated? Can I help with that?

**Comment by mrn-mln** [comment]
hey @mats-claassen, I'm going to complete face-mask at this repo for the next 2-3 days. Please let me know if you wanna contribute to this task. I plan to add a new face detector; I would be pleased if you join.

File: libs/classifiers/edgetpu/face_mask.py [hunk 5]
```diff
@@ -19,29 +19,13 @@ class Classifier:
 
     def __init__(self, config):
         self.config = config
-        # Get the model name from the config
-        self.model_name = self.config.get_section_dict('Classifier')['Name']
-        # Frames Per Second
+        self.model_name = "OFMClassifier_edgetpu.tflite"
+        self.model_path = 'libs/classifiers/edgetpu/' + self.model_name
```
Comment (mats-claassen) [inline libs/classifiers/edgetpu/face_mask.py:23]: I think it is better to store the models under `/repo/data/...` as then they are on the volume and only need to be downloaded once. Unless there is another reason they have to be in the libs folder

File: libs/classifiers/edgetpu/face_mask.py [hunk 6]
```diff
@@ -50,35 +34,35 @@ def __init__(self, config):
         self.input_details = self.interpreter.get_input_details()
         self.output_details = self.interpreter.get_output_details()
 
-
-    def inference(self, resized_rgb_image):
+    def inference(self, resized_rgb_image) -> list:
```
Comment (mats-claassen) [inline libs/classifiers/edgetpu/face_mask.py:37]: If this parameter is an array why don't we change the name to `resized_rgb_images` or something like that?

File: libs/classifiers/x86/face_mask.py [hunk 7]
```diff
@@ -1,72 +1,72 @@
...
     from neuralet repository automatically.
-
     :param config: Is a ConfigEngine instance which provides necessary parameters.
     """
 
     def __init__(self, config):
         self.config = config
-        self.model_name = self.config.get_section_dict('Classifier')['Name']
-        self.classifier_model = load_model(self.model_name)
+        self.model_path = self.config.get_section_dict('Classifier')['ModelPath']
```
Comment (mats-claassen) [inline libs/classifiers/x86/face_mask.py:19]: I think we should handle this the same way for all platforms, for example EdgeTPU/Coral don't use this config I think

File: libs/classifiers/x86/face_mask.py [hunk 8]
```diff
@@ -1,72 +1,72 @@
...
+
+        if len(self.model_path) > 0:
+            print('using %s as model' % self.model_path)
+        else:
+            url = 'https://github.com/neuralet/neuralet-models/raw/master/amd64/OFMClassifier/OFMClassifier.h5'
+            model_file = 'OFMClassifier.h5'
+            model_dir = 'data'
```
Comment (mats-claassen) [inline libs/classifiers/x86/face_mask.py:26]: Is this the same folder as /repo/data?

File: libs/classifiers/edgetpu/face_mask.py [hunk 9]
```diff
@@ -50,35 +34,35 @@ def __init__(self, config):
...
+            inference_time = time.perf_counter() - t_begin  # Second
+            self.fps = convert_infr_time_to_fps(inference_time)
+            net_output = self.interpreter.get_tensor(self.output_details[0]['index'])[0]
+            net_results.append(net_output)
+            result.append(np.argmax(net_output))  # returns class id
+
+        # TODO: optimized without for
+        scores = []
+        for i, itm in enumerate(net_results):
+            scores.append(1)
```
Comment (mats-claassen) [inline libs/classifiers/edgetpu/face_mask.py:66]: Is it correct that scores is an array of 1s?

**Comment by mats-claassen** [comment]
Hey @mrn-mln just reviewed your latest changes. Is this ready to test? Do you still want to add models? What about the Jetson's? Maybe I can help with some of this

File: libs/classifiers/edgetpu/face_mask.py (hunk 6 above)
Comment (mrn-mln) [reply libs/classifiers/edgetpu/face_mask.py:37]: It's an array of resized images. I changed its name to resized_rgb_images

File: libs/classifiers/x86/face_mask.py (hunk 8 above)
Comment (mrn-mln) [reply libs/classifiers/x86/face_mask.py:26]: Yes, I modified it to /repo/data/...

File: libs/classifiers/edgetpu/face_mask.py (hunk 9 above)
Comment (mrn-mln) [reply libs/classifiers/edgetpu/face_mask.py:66]: @alpha-carinae29

**Comment by mrn-mln** [comment]
@mats-claassen thank you for reviewing, I applied your comments to the code.
I tested it on x86, for Jetson, first of all, we need to deploy openpifpaf after that I will implement Jetson's.
Do you have coral device? I plan to quantize [this model](https://github.com/peiyunh/tiny) as a face detector on edgetpu devices

**Comment by mats-claassen** [comment]
@mrn-mln I currently do not have a Coral Dev board but should have one next week. I do have a Jetson.
As you mention that model, I see it has a non-commercial research use only license. I also saw that in the Neuralet PR you added a specific license for Facemask. We should add it here as well, I think

**Comment by mrn-mln** [comment]
@mats-claassen yes I will add the license. Thanks

File: libs/classifiers/edgetpu/face_mask.py (hunk 9 above)
Comment (mrn-mln) [reply libs/classifiers/edgetpu/face_mask.py]: @alpha-carinae29, I fixed it.

**Comment by mrn-mln** [comment]
@mats-claassen thanks for reviewing, @mhejrati the PR is tested on Coral Dev Board, Coral USB Accelerator, and x86 successfully. It's ready for merging.
I will send another PR for jetson after deploying openpifpaf on a such device.

**Comment by mats-claassen** [comment]
It would also be good to add the `Classifier` section to the Openvino config file. I guess the OFMClassifier also works on Openvino?

**Pull request merged**